### PR TITLE
Isen dev2

### PR DESCRIPTION
**Pull request by isentropic**
Refactor of  #2567.
Please view the comments in the code.
Now ,
```
scatter(rand(5), rand(5), markerstrokealpha = .5, m=:star,cmarkersize=30, markerstrokewidth=2, thickness_scaling=1, markeralpha=.3)
```
produces:

Figure_3

That is, `markerstrokealpha, markeralpha` work as intended.

We also need to adjust marker size in the legend, in case `thickness_scaling` is used. However, I could not find a good way to this

### CODE REVIEW

File: src/backends/pyplot.jl [hunk 1]
```diff
@@ -229,11 +229,6 @@ function fix_xy_lengths!(plt::Plot{PyPlotBackend}, series::Series)
     end
 end
 
-py_linecolor(series::Series)          = py_color(series[:linecolor])
-py_markercolor(series::Series)        = py_color(series[:markercolor])
-py_markerstrokecolor(series::Series)  = py_color(series[:markerstrokecolor])
-py_fillcolor(series::Series)          = py_color(series[:fillcolor])
-
 py_linecolormap(series::Series)       = py_colormap(series[:linecolor])
```
Comment (isentropic) [inline src/backends/pyplot.jl:232]: I could not understand how `py_linecolormap` should be replaced

File: src/backends/pyplot.jl [hunk 2]
```diff
@@ -533,19 +528,19 @@ function py_add_series(plt::Plot{PyPlotBackend}, series::Series)
...
+                    extrakw[:c] = [py_color(get_markercolor(series, i), get_markercoloralpha(series, i))]
                 end
 
                 push!(handle, ax."scatter"(_cycle(x,i), _cycle(y,i);
                     label = series[:label],
                     zorder = series[:series_plotindex] + 0.5,
                     marker = py_marker(_cycle(shapes,i)),
                     s =  py_thickness_scale(plt, _cycle(series[:markersize],i) .^ 2),
-                    facecolors = get_markercolor(series, i),
+                    facecolors = get_markercolor(series, i),  # Not sure if we need  get_markercoloralpha(series, i) here
```
Comment (isentropic) [inline src/backends/pyplot.jl:543]: I'm uncertain if this is right

Comment (isentropic) [reply src/backends/pyplot.jl:543]: Should I also add `markercoloralpha`?

File: src/backends/pyplot.jl [hunk 3]
```diff
@@ -1320,7 +1315,7 @@ function py_add_legend(plt::Plot, sp::Subplot, ax)
                         color = py_color(single_color(get_linecolor(series, clims)), get_linealpha(series)),
                         linewidth = py_thickness_scale(plt, clamp(get_linewidth(series), 0, 5)),
                         linestyle = py_linestyle(:path, get_linestyle(series)),
-                        marker = py_marker(_cycle(series[:markershape], 1)),
+                        marker = py_marker(_cycle(series[:markershape], 1)),  # Need to adjust markersize to match the legend box size
```
Comment (isentropic) [inline src/backends/pyplot.jl:1318]: Here, we can specify `markersize` too, but it only needs to trigger when `thickness_scaling` is used.

File: src/backends/pyplot.jl (hunk 2 above)
Comment (daschw) [reply src/backends/pyplot.jl:543]: Yes

File: src/backends/pyplot.jl (hunk 1 above)
Comment (daschw) [reply src/backends/pyplot.jl:232]: Yeah, leave that as it is.

**Comment by isentropic** [comment]
Should be good now

**Comment by isentropic** [comment]
Relevant issue https://github.com/JuliaPlots/Plots.jl/issues/2447

**Comment by daschw** [comment]
Nice, it would be great if you could test locally that everything works as expected using these lines in test/runtests.jl: https://github.com/JuliaPlots/Plots.jl/blob/master/test/runtests.jl#L60-L63

**Comment by isentropic** [comment]
For some reason I can't even pass the tests before my changes.
```
Test Summary: | Pass  Total
HDF5_Plots    |    2      2
Test Summary: | Pass  Total
PGFPlotsX     |   41     41
[ Info: Testing plot: pyplot:1:Lines
PyPlot: Error During Test at /home/zhanibek/Desktop/Software/Plots.jl/test/imgcomp.jl:40
```

**Comment by daschw** [comment]
What error do you get?

**Comment by isentropic** [comment]
> What error do you get?

```
yPlot: Error During Test at /home/zhanibek/Desktop/Software/Plots.jl/test/imgcomp.jl:40
  Test threw exception
  Expression: image_comparison_tests(pkg, i, debug = debug, sigma = sigma, tol = tol) |> success == true
  MethodError: no method matching _show(::IOStream, ::MIME{Symbol("application/pdf")}, ::Plots.Plot{Plots.PyPlotBackend})
  The applicable method may be too new: running in world age 27472, while current world is 27562.
  Closest candidates are:
    _show(::IO, ::MIME{Symbol("application/pdf")}, ::Plots.Plot{Plots.PyPlotBackend}) at /home/zhanibek/Desktop/Software/Plots.jl/src/backends/pyplot.jl:1411 (method too new to be called from this world context.)
    _show(::IO, !Matched::MIME{Symbol("text/html")}, ::Plots.Plot) at /home/zhanibek/Desktop/Software/Plots.jl/src/output.jl:181
    _show(::IO, !Matched::MIME{Symbol("text/plain")}, ::Plots.Plot) at /home/zhanibek/Desktop/Software/Plots.jl/src/output.jl:225
    ...
  Stacktrace:
   [1] show(::IOStream, ::MIME{Symbol("application/pdf")}, ::Plots.Plot{Plots.PyPlotBackend}) at /home/zhanibek/Desktop/Software/Plots.jl/src/output.jl:216
   [2] pdf(::Plots.Plot{Plots.PyPlotBackend}, ::String) at /home/zhanibek/Desktop/Software/Plots.jl/src/output.jl:25
   [3] _show_pdfbackends(::IOStream, ::MIME{Symbol("image/png")}, ::Plots.Plot{Plots.PyPlotBackend}) at /home/zhanibek/Desktop/Software/Plots.jl/src/fileio.jl:11
   [4] _show(::IOStream, ::MIME{Symbol("image/png")}, ::Plots.Plot{Plots.PyPlotBackend}) at /home/zhanibek/Desktop/Software/Plots.jl/src/init.jl:97
   [5] show(::IOStream, ::MIME{Symbol("image/png")}, ::Plots.Plot{Plots.PyPlotBackend}) at /home/zhanibek/Desktop/Software/Plots.jl/src/output.jl:216
   [6] png(::Plots.Plot{Plots.PyPlotBackend}, ::String) at /home/zhanibek/Desktop/Software/Plots.jl/src/output.jl:8
   [7] png(::String) at /home/zhanibek/Desktop/Software/Plots.jl/src/output.jl:11
   [8] (::var"#75#76"{Plots.PlotExample})(::String, ::Int64) at /home/zhanibek/Desktop/Software/Plots.jl/test/imgcomp.jl:23
   [9] test_images(::VisualTest; popup::Bool, kw::Base.Iterators.Pairs{Symbol,Any,Tuple{Symbol,Symbol,Symbol},NamedTuple{(:sigma, :tol, :newfn),Tuple{Array{Int64,1},Float64,String}}}) at /home/zhanibek/.julia/packages/VisualRegressionTests/q0Rxt/src/imgcomp.jl:48
   [10] image_comparison_tests(::Symbol, ::Int64; debug::Bool, popup::Bool, sigma::Array{Int64,1}, tol::Float64) at /home/zhanibek/Desktop/Software/Plots.jl/test/imgcomp.jl:28
   [11] image_comparison_facts(::Symbol; skip::Array{Int64,1}, only::Nothing, debug::Bool, sigma::Array{Int64,1}, tol::Float64) at /home/zhanibek/Desktop/Software/Plots.jl/test/imgcomp.jl:40
   [12] top-level scope at /home/zhanibek/Desktop/Software/Plots.jl/test/runtests.jl:49
   [13] top-level scope at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.4/Test/src/Test.jl:1113
   [14] top-level scope at /home/zhanibek/Desktop/Software/Plots.jl/test/runtests.jl:49

[ Info: Testing plot: pyplot:3:Parametric plots

signal (11): Segmentation fault
in expression starting at /home/zhanibek/Desktop/Software/Plots.jl/test/runtests.jl:48
_PyFunction_Vectorcall at /usr/lib/libpython3.8.so.1.0 (unknown line)
PyObject_Call at /usr/lib/libpython3.8.so.1.0 (unknown line)
macro expansion at /home/zhanibek/.julia/packages/PyCall/zqDXB/src/exception.jl:93 [inlined]
#110 at /home/zhanibek/.julia/packages/PyCall/zqDXB/src/pyfncall.jl:43 [inlined]
disable_sigint at ./c.jl:446 [inlined]
__pycall! at /home/zhanibek/.julia/packages/PyCall/zqDXB/src/pyfncall.jl:42 [inlined]
_pycall! at /home/zhanibek/.julia/packages/PyCall/zqDXB/src/pyfncall.jl:29
_pycall! at /home/zhanibek/.julia/packages/PyCall/zqDXB/src/pyfncall.jl:11 [inlined]
#pycall#116 at /home/zhanibek/.julia/packages/PyCall/zqDXB/src/pyfncall.jl:83 [inlined]
pycall at /home/zhanibek/.julia/packages/PyCall/zqDXB/src/pyfncall.jl:83 [inlined]
gcf at /home/zhanibek/.julia/packages/PyPlot/XHEG0/src/PyPlot.jl:154
_create_backend_figure at /home/zhanibek/Desktop/Software/Plots.jl/src/backends/pyplot.jl:335
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2144 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2322
jl_apply at /buildworker/worker/package_linux64/build/src/julia.h:1692 [inlined]
do_apply at /buildworker/worker/package_linux64/build/src/builtins.c:643
jl_f__apply_latest at /buildworker/worker/package_linux64/build/src/builtins.c:693
#invokelatest#1 at ./essentials.jl:712 [inlined]
invokelatest at ./essentials.jl:711 [inlined]
_plot_setup at /home/zhanibek/Desktop/Software/Plots.jl/src/pipeline.jl:170
plot_setup! at /home/zhanibek/Desktop/Software/Plots.jl/src/pipeline.jl:156 [inlined]
recipe_pipeline! at /home/zhanibek/.julia/packages/RecipesPipeline/NRqoc/src/RecipesPipeline.jl:84
_plot! at /home/zhanibek/Desktop/Software/Plots.jl/src/plot.jl:167
unknown function (ip: 0x7f32d9d4aa7e)
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2158 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2322
#plot#122 at /home/zhanibek/Desktop/Software/Plots.jl/src/plot.jl:57
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2158 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2322
jl_apply at /buildworker/worker/package_linux64/build/src/julia.h:1692 [inlined]
do_apply at /buildworker/worker/package_linux64/build/src/builtins.c:643
plot##kw at /home/zhanibek/Desktop/Software/Plots.jl/src/plot.jl:51
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2158 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2322
jl_apply at /buildworker/worker/package_linux64/build/src/julia.h:1692 [inlined]
do_call at /buildworker/worker/package_linux64/build/src/interpreter.c:369
eval_value at /buildworker/worker/package_linux64/build/src/interpreter.c:458
eval_stmt_value at /buildworker/worker/package_linux64/build/src/interpreter.c:409 [inlined]
eval_body at /buildworker/worker/package_linux64/build/src/interpreter.c:817
jl_interpret_toplevel_thunk at /buildworker/worker/package_linux64/build/src/interpreter.c:911
jl_toplevel_eval_flex at /buildworker/worker/package_linux64/build/src/toplevel.c:814
jl_toplevel_eval_in at /buildworker/worker/package_linux64/build/src/toplevel.c:843
eval at ./boot.jl:331 [inlined]
eval at ./client.jl:449
#75 at /home/zhanibek/Desktop/Software/Plots.jl/test/imgcomp.jl:22
unknown function (ip: 0x7f33016e8339)
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2144 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2322
jl_apply at /buildworker/worker/package_linux64/build/src/julia.h:1692 [inlined]
do_apply at /buildworker/worker/package_linux64/build/src/builtins.c:643
#test_images#4 at /home/zhanibek/.julia/packages/VisualRegressionTests/q0Rxt/src/imgcomp.jl:48
test_images##kw at /home/zhanibek/.julia/packages/VisualRegressionTests/q0Rxt/src/imgcomp.jl:48 [inlined]
#image_comparison_tests#74 at /home/zhanibek/Desktop/Software/Plots.jl/test/imgcomp.jl:28
image_comparison_tests##kw at /home/zhanibek/Desktop/Software/Plots.jl/test/imgcomp.jl:4 [inlined]
#image_comparison_facts#77 at /home/zhanibek/Desktop/Software/Plots.jl/test/imgcomp.jl:40
image_comparison_facts##kw at /home/zhanibek/Desktop/Software/Plots.jl/test/imgcomp.jl:37
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2158 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2322
jl_apply at /buildworker/worker/package_linux64/build/src/julia.h:1692 [inlined]
do_call at /buildworker/worker/package_linux64/build/src/interpreter.c:369
eval_value at /buildworker/worker/package_linux64/build/src/interpreter.c:458
eval_stmt_value at /buildworker/worker/package_linux64/build/src/interpreter.c:409 [inlined]
eval_body at /buildworker/worker/package_linux64/build/src/interpreter.c:817
eval_body at /buildworker/worker/package_linux64/build/src/interpreter.c:744
eval_body at /buildworker/worker/package_linux64/build/src/interpreter.c:744
jl_interpret_toplevel_thunk at /buildworker/worker/package_linux64/build/src/interpreter.c:911
jl_toplevel_eval_flex at /buildworker/worker/package_linux64/build/src/toplevel.c:814
jl_parse_eval_all at /buildworker/worker/package_linux64/build/src/ast.c:872
jl_load at /buildworker/worker/package_linux64/build/src/toplevel.c:872
include at ./client.jl:439
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2144 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2322
jl_apply at /buildworker/worker/package_linux64/build/src/julia.h:1692 [inlined]
do_call at /buildworker/worker/package_linux64/build/src/interpreter.c:369
eval_value at /buildworker/worker/package_linux64/build/src/interpreter.c:458
eval_stmt_value at /buildworker/worker/package_linux64/build/src/interpreter.c:409 [inlined]
eval_body at /buildworker/worker/package_linux64/build/src/interpreter.c:817
jl_interpret_toplevel_thunk at /buildworker/worker/package_linux64/build/src/interpreter.c:911
jl_toplevel_eval_flex at /buildworker/worker/package_linux64/build/src/toplevel.c:814
jl_toplevel_eval_flex at /buildworker/worker/package_linux64/build/src/toplevel.c:764
jl_toplevel_eval_in at /buildworker/worker/package_linux64/build/src/toplevel.c:843
eval at ./boot.jl:331
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2144 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2322
exec_options at ./client.jl:264
_start at ./client.jl:484
jfptr__start_2076.clone_1 at /home/zhanibek/julias/julia-1.4.0/lib/julia/sys.so (unknown line)
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2144 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2322
jl_apply at /buildworker/worker/package_linux64/build/ui/../src/julia.h:1692 [inlined]
true_main at /buildworker/worker/package_linux64/build/ui/repl.c:96
main at /buildworker/worker/package_linux64/build/ui/repl.c:217
__libc_start_main at /usr/lib/libc.so.6 (unknown line)
_start at /home/zhanibek/julias/julia-1.4.0/bin/julia (unknown line)
Allocations: 200202340 (Pool: 200156516; Big: 45824); GC: 187
ERROR: Package Plots errored during testing
```

**Comment by daschw** [comment]
So, here's my set-up for testing locally with backends other than GR (they are not tested on travis and hence not in the test dependencies.
I have all the packages I work on in `~/.julia/dev/`. (I don't think that is required for tests to run though). In my local clone of Plots I have a folder named `dev/` that I use to keep a separate Project.toml for testing. In this project I have all of Plots test dependencies and different backens:
```julia
(@v1.4) pkg> activate dev
 Activating environment at `~/.julia/dev/Plots/dev/Project.toml`

(dev) pkg> st
Status `~/.julia/dev/Plots/dev/Project.toml`
  [a93c6f00] DataFrames v0.20.2
  [5789e2e9] FileIO v1.2.4
  [4d00f742] GeometryTypes v0.8.2
  [4c0ca9eb] Gtk v1.1.3
  [f67ccb44] HDF5 v0.13.1
  [82e4d734] ImageIO v0.1.1
  [6218d12a] ImageMagick v1.1.4
  [916415d5] Images v0.22.1
  [47be7bcc] ORCA v0.3.1
  [6fe1bfb0] OffsetArrays v1.0.4
  [3b7a836e] PGFPlots v3.2.1
  [8314cec4] PGFPlotsX v1.2.6
  [f0f68f2c] PlotlyJS v0.13.1
  [91a5bcdd] Plots v1.0.10 [`~/.julia/dev/Plots`]
  [d330b81b] PyPlot v2.9.0
  [ce6b1742] RDatasets v0.6.7
  [aa65fe97] SnoopCompile v1.2.3
  [90137ffa] StaticArrays v0.12.1
  [f3b207a7] StatsPlots v0.14.5
  [b8865327] UnicodePlots v1.2.0
  [34922c18] VisualRegressionTests v0.3.3
  [76f85450] LibGit2 
  [9a3f8284] Random 
  [8dfed614] Test 
```
When I test a backend I do not use `test Plots` but `include("test/runtests.jl")` from this environment.
Or I just go through a few lines that I want to check in runtests.jl in Atom.

**Comment by daschw** [comment]
I just tested locally and it looks good. Thanks a lot for this!

**Pull request merged**